### PR TITLE
Hsdo 325

### PR DESCRIPTION
**Pull request by sherakama**
# Ready For Review - Do not merge

Multiple entity importing with entity references functionality.

Review:
1. Read over code changes for any glaring issues or poor choices

To Test: (Using JSA)
1. Install and connect to API
2. Create an importer and mapper for stanford_person (Use BUWS org and susskind sunet)
3. Create an importer and mapper for stanford_publication (same options as above)
4. On the publication mapper page for the entity reference field select the person importer
5. Ensure that both importers import content and that the entity reference is set
6. Delete and uninstall everything.

For the publications mapper you can use the following configuration:
screen shot 2016-05-24 at 5 21 50 pm

### CODE REVIEW

**Comment by sherakama** [comment]
@jbickar,
Please provide an extra set of eyeballs on this one as it is a large and complicated PR.

**Comment by jbickar** [comment]
👀

**Comment by zchandler** [comment]
Wow. This sounds pretty goddam amazing. Thanks for even attempting this @sherakama !!

**Comment by jbickar** [comment]
To test:
0. Check out this branch

**Comment by jbickar** [comment]
Getting AJAX 500 errors on batch import. Will test more tomorrow.

**Comment by jbickar** [comment]
To test:
-1. Check out this branch
0. Run `drush updb`

**Comment by jbickar** [comment]
Manual testing completed successfully.

File: includes/CAPx/Drupal/Importer/Orphans/EntityImporterOrphans.php [hunk 1]
```diff
@@ -331,6 +337,45 @@ public function getResults() {
...
+      $entity = entity_load($entityType, array($entityId));
+      $profile = entity_metadata_wrapper($entityType, $entity[$entityId]);
+
+      switch ($action) {
+        case 'delete':
+          $profile->delete();
+          break;
+
+        case 'block':
+        case 'unpublish':
```
Comment (jbickar) [inline includes/CAPx/Drupal/Importer/Orphans/EntityImporterOrphans.php:364]: Do you not want to `break;` out of this here?

File: includes/CAPx/Drupal/Importer/Orphans/Lookups/LookupMissingMultipleByGUUID.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    return $orphans;
+  }
+
+  /**
+   * @param $profileIDs
+   * @param $importerMachineName
+   */
+  protected function getLocalGUUIds($profileIDs, $importerMachineName) {
```
Comment (jbickar) [inline includes/CAPx/Drupal/Importer/Orphans/Lookups/LookupMissingMultipleByGUUID.php]: Is this method used anywhere?

File: includes/CAPx/Drupal/Mapper/EntityMapper.php [hunk 3]
```diff
@@ -263,6 +296,56 @@ public function mapFieldCollections($data) {
...
 
   /**
+   * Process entity reference fields uniquely.
+   *
+   * Reference fields are a special field and need to be handled differently.
+   * Allow for the ability to look for an item to attach to from the API
+   * that has already been imported.
+   *
+   */
+  public function mapReferences() {
```
Comment (jbickar) [inline includes/CAPx/Drupal/Mapper/EntityMapper.php:306]: From what I can tell, `entityreference` is not a dependency of `stanford_capx`. What happens here if `entityreference` is not enabled?

File: includes/CAPx/Drupal/Mapper/MapperAbstract.php [hunk 4]
```diff
@@ -52,10 +51,93 @@ public function getRemoteDataByJsonPath($data, $path) {
       throw new \Exception("Path cannot be empty", 1);
     }
 
+    // Wildcard paths need to have index association maintained. JSONPath does
+    // not provide much help in this and we need to do funny things.
```
Comment (jbickar) [inline includes/CAPx/Drupal/Mapper/MapperAbstract.php:55]: Funny how? Like a clown?

File: includes/CAPx/Drupal/Processors/EntityProcessor.php [hunk 5]
```diff
@@ -31,8 +31,176 @@ class EntityProcessor extends ProcessorAbstract {
...
+    }
+    catch (\Exception $e) {
+      $multi = FALSE;
+    }
+
+    // Sometimes we need to create one, other times we need moar.
+    if (!empty($multi) && $multi == 1) {
+      $entity = $this->executeMultiple();
+    }
+    else {
```
Comment (jbickar) [inline includes/CAPx/Drupal/Processors/EntityProcessor.php:49]: What happens if `$multi` is empty or `0` or a negative integer here? Does that get caught as an exception, above? Or handled in `executeSingle()`?

File: includes/CAPx/Drupal/Processors/EntityProcessor.php [hunk 6]
```diff
@@ -148,11 +317,13 @@ public function updateEntity($entity, $data, $mapper) {
    *   The data to be mapped to the new entity
    * @param object $mapper
    *   The EntityMapper instance
+   * @param mixed $guuid
+   *   The
```
Comment (jbickar) [inline includes/CAPx/Drupal/Processors/EntityProcessor.php:321]: Just gonna leave me hanging here?

**Comment by jbickar** [comment]
Comments inline.

**Comment by sherakama** [comment]
Thanks @jbickar. I'll start digesting them now.

File: includes/CAPx/Drupal/Importer/Orphans/EntityImporterOrphans.php (hunk 1 above)
Comment (sherakama) [inline includes/CAPx/Drupal/Importer/Orphans/EntityImporterOrphans.php:364]: Block and unpublish are the same action when it comes to setting the status of the entity. Block is for users and unpublish for nodes. Stacking them together here saves me writing more code.

File: includes/CAPx/Drupal/Importer/Orphans/Lookups/LookupMissingMultipleByGUUID.php (hunk 2 above)
Comment (sherakama) [inline includes/CAPx/Drupal/Importer/Orphans/Lookups/LookupMissingMultipleByGUUID.php]: Yes, line 29 above.

File: includes/CAPx/Drupal/Mapper/EntityMapper.php (hunk 3 above)
Comment (sherakama) [inline includes/CAPx/Drupal/Mapper/EntityMapper.php:306]: References act just like field collections in this case. CAPx does not require field collections either. This method is for handling the fields when they are available. When they are not available the references config setting will be an empty array.

File: includes/CAPx/Drupal/Mapper/MapperAbstract.php (hunk 4 above)
Comment (sherakama) [inline includes/CAPx/Drupal/Mapper/MapperAbstract.php:55]: Funny as in like: nervous laughter

File: includes/CAPx/Drupal/Importer/Orphans/EntityImporterOrphans.php (hunk 1 above)
Comment (jbickar) [inline includes/CAPx/Drupal/Importer/Orphans/EntityImporterOrphans.php:364]: Thanks. Code comment to that effect, then?

File: includes/CAPx/Drupal/Mapper/EntityMapper.php (hunk 3 above)
Comment (jbickar) [inline includes/CAPx/Drupal/Mapper/EntityMapper.php:306]: Great, thanks for the explanation.

File: includes/CAPx/Drupal/Processors/EntityProcessor.php (hunk 5 above)
Comment (sherakama) [inline includes/CAPx/Drupal/Processors/EntityProcessor.php:49]: Multi is a setting from a boolean checkbox. It should never be anything that other than 1's and 0's. 

In any case, I do explicitly check for the value to be == 1. If it is anything else then go to the single option where it does not use the config setting for anything.

File: includes/CAPx/Drupal/Importer/Orphans/Lookups/LookupMissingMultipleByGUUID.php (hunk 2 above)
Comment (jbickar) [inline includes/CAPx/Drupal/Importer/Orphans/Lookups/LookupMissingMultipleByGUUID.php]: WTF why did my grep only return one usage? Let me try that again.

Comment (jbickar) [inline includes/CAPx/Drupal/Importer/Orphans/Lookups/LookupMissingMultipleByGUUID.php]: Ah, you have a capitalization mismatch. `getLocalGUUIds` vs `getLocalGUUIDs` in line 29.

Comment (sherakama) [inline includes/CAPx/Drupal/Importer/Orphans/Lookups/LookupMissingMultipleByGUUID.php]: Oooo thanks! This causes big problems in systems that are sensitive to this.

**Comment by sherakama** [comment]
@boznik, 
Please review but do not merge in to 1.x

**Comment by boznik** [comment]
Running "phplint:php" (phplint) task

> > 954 files php linted.

Running "phplint:module" (phplint) task

> > 3 files php linted.

Running "phplint:inc" (phplint) task

> > 18 files php linted.

Done, without errors.

(excluding some tests:
ie: includes/vendor/phpunit/php-token-stream/tests/_fixture/class_with_method_that_declares_anonymous_class.php)

**Comment by boznik** [comment]
finding discrepancies in the api and the data browser at: capx/data-browser
ex: $.publications._.firstPublished needs to be $.publications._.firstPublished.value

File: includes/CAPx/Drupal/Importer/Orphans/EntityImporterOrphans.php [hunk 7]
```diff
@@ -432,49 +533,87 @@ public function processAdopted($profiles, $orphans) {
...
-      watchdog('EntityImporterOrphans', "%title was orphaned from the importer %importername.", array("%title" => $entity->label(), "%importername" => $importerName), WATCHDOG_NOTICE, '');
+      $keys = array(
+        'entity_type',
+        'entity_id',
+        'importer',
+        'bundle_type',
+        'guuid',
+      );
+
+      drupal_write_record('capx_profiles', $record, $keys);
```
Comment (boznik) [inline includes/CAPx/Drupal/Importer/Orphans/EntityImporterOrphans.php:611]: should we be using db_merge rather than drupal_write_record?

Comment (sherakama) [inline includes/CAPx/Drupal/Importer/Orphans/EntityImporterOrphans.php:611]: drupal_write_record with the keys does a db_update.

I just looked in to what the preferred usage should be and it looks like dwr() was not well liked and is going away in D8: https://www.drupal.org/node/1774104
https://www.drupal.org/node/2340291
https://www.drupal.org/node/2194885

I guess we should follow suit from now on. I thought dwr() was the standard but I guess it changed.

Comment (sherakama) [inline includes/CAPx/Drupal/Importer/Orphans/EntityImporterOrphans.php:611]: I'm not seeing a discrepancy in the data browser:
Shows the value

Comment (boznik) [inline includes/CAPx/Drupal/Importer/Orphans/EntityImporterOrphans.php:611]: yes, seeing the correct structure in the data browser now. not sure what was going on with that.

**Comment by boznik** [comment]
Not merging, but this code seems solid.

**Comment by tsui1** [comment]
Thanks @sherakama :unamused: :baby:

**Pull request closed without merging**